### PR TITLE
Generating IObservable type response

### PR DESCRIPTION
**Pull request by janfolbrecht**
Refit supports also IObservable reactive response type in interfaces so I added this feature to Refitter because it was missing. 
I added property ReturnIApiResponse to Settings and RefitGeneratorSettings
It replaces System.Threading.Task import by System.Reactive and it generates IObservable<Unit> instead of Task and IObservable<> instead of Task<>. 
I added one unit test to test generated output is compilable. I had to add Unit class manually to mock missing System.Reactive.Unit class.

### CODE REVIEW

File: src/Refitter.Tests/SwaggerPetstoreTests.cs [hunk 1]
```diff
@@ -414,6 +414,25 @@ public async Task Can_Build_Generated_Code_With_IApiResponse(SampleOpenSpecifica
...
 
+    [Theory]
+    [InlineData(SampleOpenSpecifications.SwaggerPetstoreJsonV3, "SwaggerPetstore.json")]
+    public async Task Can_Build_Generated_Code_With_IObservableResponse(SampleOpenSpecifications version, string filename)
+    {
+        var settings = new RefitGeneratorSettings();
+        settings.ReturnIObservable = true;
+        var generateCode = await GenerateCode(version, filename, settings);
+        //cannot build without it because System.Reactive package has to be installed first
+        generateCode += @"
```
Comment (christianhelle) [inline src/Refitter.Tests/SwaggerPetstoreTests.cs:425]: You can add the `System.Reactive` package reference to the [ProjectFileContents.cs](https://github.com/christianhelle/refitter/blob/main/src/Refitter.Tests/Build/ProjectFileContents.cs) file

```xml
<PackageReference Include=""System.Reactive"" Version=""6.0.0"" />
```

so you don't need to insert a fake `Unit` type

**Comment by christianhelle** [comment]
@all-contributors please add @janfolbrecht for ideas and code

File: src/Refitter.Core/RefitInterfaceImports.cs [hunk 2]
```diff
@@ -4,41 +4,37 @@ namespace Refitter.Core;
...
+    {
+        var namespaces = new List<string>(defaultNamespases);
+        if (settings.UseCancellationTokens)
+        {
+            namespaces.Add("System.Threading");
+        }
+
+        if (settings.ReturnIObservable)
+        {
+            namespaces.Add("System.Reactive");
```
Comment (christianhelle) [inline src/Refitter.Core/RefitInterfaceImports.cs:23]: I'm not sure I fully understand why the generated code needs a dependency on `System.Reactive`. What do you need the `Unit` type for?

File: src/Refitter.Core/RefitInterfaceGenerator.cs [hunk 3]
```diff
@@ -193,11 +196,17 @@ protected static bool IsApiResponseType(string typeName)
...
         return settings.ReturnIApiResponse
-            ? $"Task<IApiResponse<{WellKnownNamesspaces.TrimImportedNamespaces(returnTypeParameter)}>>"
-            : $"Task<{WellKnownNamesspaces.TrimImportedNamespaces(returnTypeParameter)}>";
+            ? $"{asyncType}<IApiResponse<{WellKnownNamesspaces.TrimImportedNamespaces(returnTypeParameter)}>>"
+            : $"{asyncType}<{WellKnownNamesspaces.TrimImportedNamespaces(returnTypeParameter)}>";
     }
 
+    private string GetAsyncOperationType(bool withVoidReturnType) =>
+        settings.ReturnIObservable
+            ? "IObservable" + (withVoidReturnType ? "<Unit>" : string.Empty)
```
Comment (christianhelle) [inline src/Refitter.Core/RefitInterfaceGenerator.cs:207]: I don't have much experience working with `IObservable<T>` but why do we need `Unit`? What does it do and what is it for?

The examples in the [Refit README](https://github.com/reactiveui/refit?tab=readme-ov-file#retrieving-the-response) uses an `HttpResponseMessage` together with `IObservable`

```csharp
// Returns the raw response, as an IObservable that can be used with the
// Reactive Extensions
[Get("/users/{user}")]
IObservable<HttpResponseMessage> GetUser(string user);
```

**Comment by janfolbrecht** [comment]
Hi,
yes System.Reactive is for Unit class. It is reactive "void" replacement.
I did not use HttpResponseMessage return type with refit yet, so I don't
know if it works as replacement of void or not.
Jan


st 21. 2. 2024 v 22:08 odesílatel Christian Helle ***@***.***>
napsal:

File: src/Refitter.Core/RefitInterfaceImports.cs (hunk 2 above)
Comment (janfolbrecht) [reply src/Refitter.Core/RefitInterfaceImports.cs:23]: Yes. System.Reactive is for Unit class. It is reactive "void" replacement.

**Comment by christianhelle** [comment]
@janfolbrecht are you up for fixing the things I suggested? I perfectly understand if don't feel like it and since the requested changes are trivial I can easily do this myself so we can get this merged

**Comment by janfolbrecht** [comment]
Thank you. If you can make suggested changes please do it. Jan

**Pull request merged**